### PR TITLE
[wasm][wasi] Allow to build wasi in library mode

### LINKED ISSUE #96869: [wasm][wasi] No main function should be required when building a library
### Description

Currently mono requires a main function when building for wasi. This shouldn't be required.

Related: https://github.com/dotnet/runtime/issues/96419

@AaronRobinsonMSFT @lewing @yowl @jsturtevant

### Reproduction Steps

Try to build example with `main` removed.
Example: https://github.com/silesmo/wasm-unmanaged-callers-bug/blob/main/mono-example/MonoEntrypoint.cs

### Expected behavior

No main function should be required for a library.

### Actual behavior

Main function is required.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by mkhamoyan**
By this PR we are supporting library mode for wasi.

Here are all possibilites to build wasi.

1. Create exe
      ```
      <Project Sdk="Microsoft.NET.Sdk">
        <PropertyGroup>
          <OutputType>Exe</OutputType>
          ...
        </PropertyGroup>
      </Project>
      ```
3. Use wasi specific C# api in standard library
      ```
      <Project Sdk="Microsoft.NET.Sdk">
        <PropertyGroup>
          <OutputType>Library</OutputType>
          <RuntimeIdentifier>wasi-wasm</RuntimeIdentifier>
          ...
        </PropertyGroup>
       </Project>
      ```
5. Create Wasm library
      ```
        <Project Sdk="Microsoft.NET.Sdk.WebAssembly">
          <PropertyGroup>
            <OutputType>Library</OutputType>
            <RuntimeIdentifier>wasi-wasm</RuntimeIdentifier>
            ...
          </PropertyGroup>
         </Project>
      ```

Fixes https://github.com/dotnet/runtime/issues/96869

### CODE REVIEW

File: src/mono/wasm/build/WasmApp.Common.targets [hunk 1]
```diff
@@ -173,8 +173,7 @@
     <EnableDefaultWasmAssembliesToBundle Condition="'$(EnableDefaultWasmAssembliesToBundle)' == ''">true</EnableDefaultWasmAssembliesToBundle>
     <!-- VS uses DeployOnBuild, and sdk sets _IsPublishing -->
     <WasmBuildOnlyAfterPublish Condition="'$(WasmBuildOnlyAfterPublish)' == '' and ('$(DeployOnBuild)' == 'true' or '$(_IsPublishing)' == 'true')">true</WasmBuildOnlyAfterPublish>
-    <WasmGenerateAppBundle Condition="'$(WasmGenerateAppBundle)' == '' and '$(OutputType)' != 'Library'">true</WasmGenerateAppBundle>
-    <WasmGenerateAppBundle Condition="'$(WasmGenerateAppBundle)' == ''">false</WasmGenerateAppBundle>
+    <WasmGenerateAppBundle Condition="'$(WasmGenerateAppBundle)' == ''">true</WasmGenerateAppBundle>
```
Comment (maraf) [inline src/mono/wasm/build/WasmApp.Common.targets:176]: This needs a condition as well, doesn't it? (so that the target doesn't run until it's a wasm library)

File: src/mono/sample/wasi/native/Wasi.Native.Sample.csproj [hunk 2]
```diff
@@ -8,6 +8,11 @@
     <WasmNativeStrip>false</WasmNativeStrip>
     <IsBrowserWasmProject>false</IsBrowserWasmProject>
     <WasmSingleFileBundle>true</WasmSingleFileBundle>
+    <OutputType>Library</OutputType>
+    <!-- This is needed for ReferenceCopyLocalPaths to include RuntimePackAsset-->
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
+    <!-- This is needed so it will not complain about WasmAssembliesToBundle is empty -->
+    <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
```
Comment (maraf) [inline src/mono/sample/wasi/native/Wasi.Native.Sample.csproj:15]: If the approach with SDK is working, we can move these defaults to https://github.com/dotnet/runtime/blob/main/src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Pack.props#L4

File: src/mono/wasm/build/WasmApp.Common.targets (hunk 1 above)
Comment (mkhamoyan) [reply src/mono/wasm/build/WasmApp.Common.targets:176]: shouldn't it run even when exe ?

Comment (maraf) [reply src/mono/wasm/build/WasmApp.Common.targets:176]: Yeah, but until RID=wasm+Output type=exe meant do wasm build, which we need to leave intact until wasmconsole template and runtime tests migrate to Wasm SDK. For library mode we also need to check for Wasm SDK use. (That's my theory, maybe something else guards this)

File: src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs [hunk 3]
```diff
@@ -379,6 +379,13 @@ private void EmitNativeToInterp(StreamWriter w, List<PInvokeCallback> callbacks)
             if (!is_void)
                 sb.Append($"  {MapType(method.ReturnType)} res;\n");
 
+            if (HasAttribute(method, "System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute"))
+            {
+                sb.Append($"  #if defined(WASM_LIBRARY_MODE) \n");
```
[the remark this answers is not in the document]
Comment (mkhamoyan) [reply src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs]: Implemented it by passing `IsLibraryMode` as a parameter in `ManagedToNativeGenerator`. Not sure if there is easier way to get this property there.

**Comment by ilonatommy** [comment]
Doesn't it fix https://github.com/dotnet/runtime/issues/89577 as well?

File: src/mono/nuget/Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk/Sdk/Sdk.props [hunk 4]
```diff
@@ -3,7 +3,7 @@
     <TargetArchitecture>wasm</TargetArchitecture>
     <TargetOS>wasi</TargetOS>
     <UseMonoRuntime>true</UseMonoRuntime>
-    <OutputType>Exe</OutputType>
+    <!-- <OutputType>Exe</OutputType> -->
```
Comment (maraf) [inline src/mono/nuget/Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk/Sdk/Sdk.props]: It should be fine to leave the default here. The user's csproj is imported after this file and thus it will be overriden

File: src/mono/sample/wasi/Directory.Build.targets [hunk 5]
```diff
@@ -15,13 +15,13 @@
       TODOWASI - - tcplisten localhost:64000 - - env DEBUGGER_FD=4
     -->
     <Exec WorkingDirectory="bin/wasi-wasm/AppBundle"
-          Condition="'$(WasmBuildNative)' != 'true'"
+          Condition="'$(WasmBuildNative)' != 'true' and '$(OutputType)' != 'Library'"
```
Comment (maraf) [inline src/mono/sample/wasi/Directory.Build.targets:18]: Are these changes needed?

File: src/mono/wasi/Wasi.Build.Tests/WasiLibraryModeTests.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+                        Console.WriteLine("WASM Library MyCallback is called");
+                        return 100;
+                    }
+                }
+                """;
+        string csprojCode = 
+                """
+                <Project Sdk="Microsoft.NET.Sdk.WebAssembly">
+                    <PropertyGroup>
+                        <TargetFramework>net9.0</TargetFramework>
```
Comment (maraf) [inline src/mono/wasi/Wasi.Build.Tests/WasiLibraryModeTests.cs:46]: Setting `TFM`, `RID`, `AllowUnsafeBlocks` and `WasmSingleFileBundle` (if we want to test single file bundling) should be enough

File: src/mono/wasm/build/WasmApp.Common.targets [hunk 7]
```diff
@@ -174,6 +175,7 @@
     <!-- VS uses DeployOnBuild, and sdk sets _IsPublishing -->
     <WasmBuildOnlyAfterPublish Condition="'$(WasmBuildOnlyAfterPublish)' == '' and ('$(DeployOnBuild)' == 'true' or '$(_IsPublishing)' == 'true')">true</WasmBuildOnlyAfterPublish>
     <WasmGenerateAppBundle Condition="'$(WasmGenerateAppBundle)' == '' and '$(OutputType)' != 'Library'">true</WasmGenerateAppBundle>
+    <WasmGenerateAppBundle Condition="'$(WasmGenerateAppBundle)' == '' and '$(IsLibraryMode)' == 'true'">true</WasmGenerateAppBundle>
```
Comment (maraf) [inline src/mono/wasm/build/WasmApp.Common.targets:178]: nit: It can be combined with the previous condition (`empty and (output not library or isLibraryMode)`)

File: src/mono/wasm/build/WasmApp.Common.targets [hunk 8]
```diff
@@ -162,8 +162,9 @@
     </WasmLinkDotNetDependsOn>
 
     <WasmDedup Condition="'$(WasmDedup)' == ''">true</WasmDedup>
-    <IsWasiProject Condition="'$(IsWasiProject)' == '' and '$(RuntimeIdentifier)' == 'wasi-wasm' and '$(OutputType)' != 'Library'">true</IsWasiProject>
-    <IsBrowserWasmProject Condition="'$(IsBrowserWasmProject)' == '' and '$(RuntimeIdentifier)' == 'browser-wasm' and '$(OutputType)' != 'Library'">true</IsBrowserWasmProject>
+    <IsLibraryMode Condition="'$(OutputType)' == 'Library' and '$(UsingMicrosoftNETSdkWebAssembly)' == 'true'">true</IsLibraryMode>
```
Comment (maraf) [inline src/mono/wasm/build/WasmApp.Common.targets:165]: Other targets (android, ios) are using a "private" property. I would align with them

```suggestion
    <_IsLibraryMode Condition="'$(OutputType)' == 'Library' and '$(UsingMicrosoftNETSdkWebAssembly)' == 'true'">true</IsLibraryMode>
```

File: src/mono/nuget/Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk/Sdk/Sdk.props (hunk 4 above)
Comment (mkhamoyan) [reply src/mono/nuget/Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk/Sdk/Sdk.props]: Actually it is overriding user's specified `OutputType`. 
I've updated this to check for `<OutputType Condition="'$(OutputType)' == ''">Exe</OutputType>`

File: src/mono/sample/wasi/Directory.Build.targets (hunk 5 above)
Comment (mkhamoyan) [reply src/mono/sample/wasi/Directory.Build.targets:18]: These I added for `Wasi.Console.Sample` testing purposes as library. Will revert it.

**Comment by lewing** [comment]
Is this waiting for other changes or ready to go?

File: src/mono/wasi/runtime/main.c [hunk 9]
```diff
@@ -10,7 +11,31 @@ const char* dotnet_wasi_getentrypointassemblyname();
...
+#ifdef WASM_LIBRARY_MODE
+// _initialize is a function generated by the WASI SDK libc that calls the LLVM synthesized __wasm_call_ctors function for reactor components:
+// https://github.com/WebAssembly/wasi-libc/blob/9f51a7102085ec6a6ced5778f0864c9af9f50000/libc-bottom-half/crt/crt1-reactor.c#L7-L27
+// We define and call it for WASM_LIBRARY_MODE and TARGET_WASI to call all the global c++ static constructors.  This ensures the runtime is initialized
+// when calling into WebAssembly Component Model components.
+extern void _initialize();
+
+// CustomNativeMain programs are built using the same libbootstrapperdll as WASM_LIBRARY_MODE but wasi-libc will not provide an _initialize implementation,
+// so create a dummy one here and make it weak to allow wasi-libc to provide the real implementation for WASI reactor components.
+__attribute__((weak)) void _initialize()
```
Comment (maraf) [inline src/mono/wasi/runtime/main.c:23]: From what I understood about `_initialize`, this is an important bit https://github.com/dotnet/runtimelab/blob/feature/NativeAOT-LLVM/src/coreclr/nativeaot/Bootstrap/main.cpp#L194. We don't support custom unmanaged main, so maybe this is not needed on Mono?

@dicej @SingleAccretion Please, do you have more details about `_initialize`?

**Comment by mkhamoyan** [comment]
> Is this waiting for other changes or ready to go?

@lewing It is ready to go unless there are further changes needed based on the review.

File: src/mono/wasi/runtime/main.c (hunk 9 above)
Comment (SingleAccretion) [reply src/mono/wasi/runtime/main.c:23]: `_initialize` is part of the WASI reactor ABI. It is similar to `DllMain` on Windows. One responsibility of `_initialize` is to run global (C++) constructors (`__attribute__ ((constructor))` methods).

The code being copied here is a workaround for the fact that global constructors are/were broken with components (https://github.com/dotnet/runtimelab/issues/2359).

The workaround as copied here is not correct, as you pointed out. The `_initialize` call needs to be guarded on `if (<global constructors have not been called yet>)`.

File: src/mono/wasi/runtime/main.c [hunk 10]
```diff
@@ -10,7 +11,31 @@ const char* dotnet_wasi_getentrypointassemblyname();
...
+
+#endif
+
+int initialize_runtime()
+{
+#if defined(WASM_LIBRARY_MODE)
+    if (runtime_initialized)
+		return 0;
+    _initialize();
+	runtime_initialized = true;
```
Comment (SingleAccretion) [inline src/mono/wasi/runtime/main.c:37]: This does not look thread safe.

(I know WASI is not MT ready in general, but it makes future work harder.)

File: src/mono/wasi/runtime/main.c (hunk 9 above)
Comment (dicej) [reply src/mono/wasi/runtime/main.c:23]: Yeah, `_initialize` should be called at most once, per https://github.com/WebAssembly/WASI/blob/main/legacy/application-abi.md and https://github.com/WebAssembly/tool-conventions/pull/203

File: src/mono/wasi/runtime/main.c (hunk 10 above)
Comment (mkhamoyan) [reply src/mono/wasi/runtime/main.c]: I agree, but currently, everything is single-threaded. We'll keep thread safety in mind for future work.

File: src/mono/wasi/runtime/main.c (hunk 9 above)
Comment (mkhamoyan) [reply src/mono/wasi/runtime/main.c:23]: Thanks, updated to call `_initialize` only once.

**Review by maraf — APPROVED** [review]
Looks good to me 👍

**Comment by mkhamoyan** [comment]
Failures are not related to PR changes.
